apps/prairielearn/src/lib/workspace.js

```diff
@@ -69,7 +91,13 @@ module.exports = {
    */
   connection(socket) {
     socket.on('joinWorkspace', (msg, callback) => {
-      const workspace_id = msg.workspace_id;
+      // The middleware will have already ensured that this property exists and
+      // that the client possesses a token that is valid for this workspace ID.
+      //
+      // TODO: once all clients include a token in the handshake, we can remove
+      // the code that reads the workspace ID from the message.
+      const workspace_id = socket.handshake.auth.workspace_id ?? msg.workspace_id;
```

Comment:
To be more specific, we'll be able to read `const workspace_id = socket.handshake.auth.workspace_id` at the top level of this `connection(...)` function, which also means we can collapse this explanatory comment into one place. This will also means we don't have to pass a `workspace_id` in every single message from the client.